### PR TITLE
various cleanup

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*: 
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-XYZ

### Changes and notes
 

I got a little sloppy and merged various changes without running the
code_reviewer. This PR addresses the suggested changes

### CODE REVIEW

**Comment by stevebachmeier** [comment]
## PR Review: Monorepo Support for vivarium_build_utils (epic/monorepo)

### Summary

This branch (7 commits, ~350 net additions) adds monorepo CI support to the Jenkins shared library. The core concept: a top-level monorepo.groovy provisions per-package Multibranch Pipeline jobs via Job DSL, each per-package pipeline reuses reusable_pipeline.groovy as before, and build_stages.groovy wraps every stage in `withWorkingDirectory()` so builds cd into the correct `libs/<pkg>` subdirectory. Supporting changes make the artifactory index URL configurable, search recursively for `py.typed` markers, add a tag prefix mechanism for per-package releases, and add an `env_reqs` parameter for specifying CI install extras.

### Design

1. **`tag-version` does not support `TAG_PREFIX` but `validate-tag` does.** base.mk hardcodes `git tag -a "v${PACKAGE_VERSION}"`, while validate_tag_version.sh supports `TAG_PREFIX` for monorepo tags like `vivarium-core-v1.2.3`. If a monorepo package uses `tag-version`, it will create unprefixed tags that fail validation. These two targets must be consistent. *(Design, Functionality)*

   **Suggestion:** Add `TAG_PREFIX ?=` and update `tag-version`:
   ```makefile
   TAG_PREFIX ?=
   tag-version:
   	git tag -a "${TAG_PREFIX}v${PACKAGE_VERSION}" -m "Tag automatically generated from Jenkins."
   	git push --tags
   ```

2. **`installPackage` runs a redundant `uv pip install` with a hardcoded artifactory URL.** build_stages.groovy runs `make install` (which uses `EXTRA_INDEX_FLAGS` via `IHME_PYPI`), then immediately runs a second `uv pip install .` with the artifactory URL hardcoded. This overwrites the editable install with a non-editable one and bypasses the new `IHME_PYPI` abstraction introduced in base.mk. *(Design, DRY)*

   **Suggestion:** Remove the second `uv pip install .` call. If a non-editable install is needed for CI, create a dedicated Make target.

3. **`get_package_subdir()` JOB_NAME parsing is position-dependent.** get_package_subdir.groovy checks `parts[2] == 'libs'`, assuming the Jenkins folder prefix is always a single segment. Changing the folder hierarchy (e.g., nesting under `Public/Monorepos/`) would silently return `''` instead of failing. *(Design, Maintainability)*

   **Suggestion:** An `indexOf('libs')` approach would be more resilient:
   ```groovy
   def libsIdx = parts.toList().indexOf('libs')
   if (libsIdx >= 0 && libsIdx + 1 < parts.size()) {
       return "${parts[libsIdx]}/${parts[libsIdx + 1]}"
   }
   return ''
   ```

4. **`IHME_PYPI` is used for both fetching and uploading packages.** base.mk (index) and base.mk (twine upload). If someone sets `IHME_PYPI=` (empty) for a non-IHME environment, `deploy-package-artifactory` will call `twine upload --repository-url` with an empty URL and fail cryptically. Consider adding a guard: `@[ "${IHME_PYPI}" ] || ( echo "IHME_PYPI is not set; cannot deploy"; exit 1 )`.

### Maintainability

1. **`'libs'` is a magic path segment with no single source of truth.** It appears in get_package_subdir.groovy, monorepo.groovy docstring, and is assumed by the Jenkinsfile paths config. A comment in get_package_subdir.groovy should establish this as the convention. *(Maintainability)*

2. **`env_reqs` default semantics span three files.** reusable_pipeline.groovy defaults to `""`, build_stages.groovy conditionally passes `ENV_REQS=...`, and base.mk defaults to `dev`. A reader of reusable_pipeline.groovy might think nothing is installed when in reality `dev` extras are. Add a comment: `// Empty string leaves base.mk's default ("dev") in effect`. *(Maintainability)*

3. **`cleanup()` sequence is correct but non-obvious.** build_stages.groovy runs `make clean` inside `withWorkingDirectory`, then `conda env remove` outside it (which is correct since `CONDA_ENV_PATH` is absolute). A one-line comment before the conda line would help: `// CONDA_ENV_PATH is absolute; no working dir needed`. *(Maintainability)*

4. **monorepo.groovy and get_package_subdir.groovy derive repo identity differently.** monorepo.groovy uses `GIT_URL`; get_package_subdir.groovy uses `JOB_NAME`. Both are correct for their contexts, but a comment in get_package_subdir.groovy noting that per-package JOB_NAMEs have a known format (because they're provisioned by monorepo.groovy) would clarify the different strategies. *(Maintainability)*

### DRY

1. **Artifactory URL is hardcoded in build_stages.groovy L109**, duplicating what `IHME_PYPI` / `EXTRA_INDEX_FLAGS` already centralizes in base.mk. If the URL changes, it must be updated in both places. *(DRY, Design - also flagged above)*

2. **Repository URL parsing differs between monorepo.groovy and multiPipelines.groovy.** monorepo.groovy uses `tokenize('/').last().replace('.git','')` while multiPipelines.groovy uses a regex. Low risk since they operate at different layers, but worth noting for consistency.

### Tests

1. **No automated test infrastructure exists** in this repository - no test directories, no Jenkins Pipeline Unit tests, no shell script tests (BATS or otherwise). This is the status quo, not introduced by this PR. *(Pre-existing)*

2. **`get_package_subdir()` parsing logic has no tests** and is the routing linchpin for all monorepo builds. Edge cases (branches with `/`, unexpected folder depths) could silently misroute. *(Tests)*

3. **validate_tag_version.sh TAG_PREFIX behavior has no tests.** The `git tag --list "${TAG_PREFIX}*"` glob could match tags with prefixes that are substrings of other prefixes (e.g., `TAG_PREFIX=core-` matching `core-utils-v1.0.0`). *(Tests)*

### Documentation

1. **No README documentation of monorepo support.** This is the biggest gap. A developer setting up a new monorepo has no high-level guidance on how to call `monorepo()`, structure Jenkinsfiles, or set `TAG_PREFIX`. *(Documentation)*

2. **get_python_versions.groovy lacks Groovydoc.** The new `subdir` parameter at get_python_versions.groovy has no documentation explaining its purpose or default behavior. *(Documentation)*

3. **CHANGELOG "Add support for monorepo" is too vague** for a change spanning 3 new files and 5 modified files. Consider being more specific about what was added (pipeline provisioning, working directory routing, tag prefix support). *(Documentation)*

4. **`TAG_PREFIX` in validate_tag_version.sh lacks a usage example.** Add something like: `# Example: TAG_PREFIX="vivarium-core-" for tags like "vivarium-core-v1.2.3"`. *(Documentation)*

5. **monorepo.groovy L47** comment says "we only provision from the default branch" but nothing visible in this file enforces that. Clarify that the calling Jenkinsfile must gate this step.  *(Documentation)*

### Functionality

1. **`tag-version` / `validate-tag` mismatch is a correctness bug.** Any monorepo package attempting to deploy via `tag-version` will create unprefixed tags that fail `validate-tag` when `TAG_PREFIX` is set. This must be fixed before monorepo deploys are attempted. *(Functionality, Design)*

2. **The second `uv pip install .` in `installPackage` silently overwrites the editable install.** `make install` does `uv pip install -e .` (editable), then the inline command does `uv pip install .` (non-editable). If any downstream stage relies on editable-install behavior (e.g., in-place code changes during tests), this could cause subtle failures. *(Functionality)*

3. **`deploy-package-artifactory` with empty `IHME_PYPI` will produce a broken `twine upload` command.** The credentials are validated but the URL is not. *(Functionality)*

4. **validate_tag_version.sh prefix glob `"${TAG_PREFIX}*"` is unanchored at the end** - a `TAG_PREFIX=core-` would match both `core-v1.0.0` and `core-utils-v1.0.0`. In practice this is unlikely given the naming convention, but a stricter pattern like `"${TAG_PREFIX}v*"` would be safer. *(Functionality)*

### Minor Nits

1. multiPipelines.groovy - `jenkinsfilePathsStr` has a `Str` suffix suggesting a string, but it's actually a `List<String>`. Consider renaming to `jenkinsfilePaths` for clarity.

2. build_stages.groovy - `withWorkingDirectory` silently defaults to `'.'` for single-repo builds. A debug-level log (`echo "Working directory: ${subdir ?: 'repo root'}"`) would help troubleshoot monorepo routing issues.

3. multiPipelines.groovy - The Jenkins credential UUID `'fad62062-b1f4-447b-997f-005d6b1ea41e'` is hardcoded. It has a helpful inline comment, but consider passing it as a parameter from monorepo.groovy for reusability across Jenkins instances.

### Overall

The monorepo architecture is sound - the provisioner / per-package separation is clean, `withWorkingDirectory` is applied consistently, and the new files are well-structured. The two most important issues to address before merging are:

1. **`tag-version` must respect `TAG_PREFIX`** - without this, monorepo deploys are broken.
2. **Remove or consolidate the hardcoded `uv pip install` in `installPackage`** - it contradicts the `IHME_PYPI` abstraction and does a redundant non-editable install.

Secondary priorities are adding a `IHME_PYPI` guard to `deploy-package-artifactory`, adding README documentation for monorepo usage, and considering a more resilient `get_package_subdir` implementation.

**Review by albrja — APPROVED** [review]

File: vars/monorepo.groovy [hunk 1]
```diff
@@ -24,28 +26,34 @@ def call(Map config = [:]) {
     String repositoryName = env.GIT_URL.tokenize('/').last().replace('.git', '')
     String folderPrefix = config.folderPrefix ?: 'Public'
     String rootFolderPath = "${folderPrefix}/${repositoryName}"
+    // Jenkins credential ID for the GitHub App on ihmeuw. Override if/when vbu
+    // is reused across orgs or Jenkins instances with different credentials.
+    String githubCredentialsId = config.githubCredentialsId ?: 'fad62062-b1f4-447b-997f-005d6b1ea41e'
```
Comment (rmudambi) [inline vars/monorepo.groovy]: Should we even have a default in VBU? Maybe just define it in the monorepo?

File: vars/monorepo.groovy [hunk 2]
```diff
@@ -24,28 +26,34 @@ def call(Map config = [:]) {
...
         jobDsl(
             scriptText: libraryResource('multiPipelines.groovy'),
             additionalParameters: [
-                jenkinsfilePathsStr: paths,
-                rootFolderStr      : rootPath,
-                repositoryURL      : repoURL
+                jenkinsfilePathStrings: paths,
+                rootFolderStr         : rootPath,
+                repositoryURL         : repoURL,
+                githubCredentialsId   : credId
```
Comment (rmudambi) [inline vars/monorepo.groovy:51]: Why do we change the name of the argument from `credId` to `githubCredentialsId`?

**Review by rmudambi — APPROVED** [review]

File: vars/monorepo.groovy (hunk 2 above)
Comment (stevebachmeier) [reply vars/monorepo.groovy:51]: 🤷. Just matching args between two different files. Same as the other parameters above it.
